internal/ls/completions.go

```diff
@@ -4860,3 +4985,621 @@ func createCompletionDetailsForSymbol(
+	switch tag.Kind {
+	case ast.KindJSDocParameterTag, ast.KindJSDocPropertyTag, ast.KindJSDocReturnTag, ast.KindJSDocTypeTag,
+		ast.KindJSDocTypedefTag, ast.KindJSDocSatisfiesTag:
+		return true
+	case ast.KindJSDocTemplateTag:
+		return tag.AsJSDocTemplateTag().Constraint != nil
+	default:
+		return false
+	}
+}
+
+func (l *LanguageService) jsDocCompletionInfo(
+	clientOptions *lsproto.CompletionClientCapabilities,
+	position int,
+	file *ast.SourceFile,
+	items []*lsproto.CompletionItem,
+) *lsproto.CompletionList {
+	defaultCommitCharacters := getDefaultCommitCharacters(false /*isNewIdentifierLocation*/)
+	itemDefaults := l.setItemDefaults(
+		clientOptions,
+		position,
+		file,
+		items,
+		&defaultCommitCharacters,
+		nil, /*optionalReplacementSpan*/
+	)
+	return &lsproto.CompletionList{
+		IsIncomplete: false,
+		ItemDefaults: itemDefaults,
+		Items:        items,
+	}
+}
+
+var jsDocTagNames = []string{
+	"abstract",
+	"access",
+	"alias",
+	"argument",
+	"async",
+	"augments",
+	"author",
+	"borrows",
+	"callback",
+	"class",
+	"classdesc",
+	"constant",
+	"constructor",
+	"constructs",
+	"copyright",
+	"default",
+	"deprecated",
+	"description",
+	"emits",
+	"enum",
+	"event",
+	"example",
+	"exports",
+	"extends",
+	"external",
+	"field",
+	"file",
+	"fileoverview",
+	"fires",
+	"function",
+	"generator",
+	"global",
+	"hideconstructor",
+	"host",
+	"ignore",
+	"implements",
+	"import",
+	"inheritdoc",
+	"inner",
+	"instance",
+	"interface",
+	"kind",
+	"lends",
+	"license",
+	"link",
+	"linkcode",
+	"linkplain",
+	"listens",
+	"member",
+	"memberof",
+	"method",
+	"mixes",
+	"module",
+	"name",
+	"namespace",
+	"overload",
+	"override",
+	"package",
+	"param",
+	"private",
+	"prop",
+	"property",
+	"protected",
+	"public",
+	"readonly",
+	"requires",
+	"returns",
+	"satisfies",
+	"see",
+	"since",
+	"static",
+	"summary",
+	"template",
+	"this",
+	"throws",
+	"todo",
+	"tutorial",
+	"type",
+	"typedef",
+	"var",
+	"variation",
+	"version",
+	"virtual",
+	"yields",
+}
+
+var jsDocTagNameCompletionItems = sync.OnceValue(func() []*lsproto.CompletionItem {
+	items := make([]*lsproto.CompletionItem, 0, len(jsDocTagNames))
+	for _, tagName := range jsDocTagNames {
+		item := &lsproto.CompletionItem{
+			Label:    tagName,
+			Kind:     ptrTo(lsproto.CompletionItemKindKeyword),
+			SortText: ptrTo(string(SortTextLocationPriority)),
+		}
+		items = append(items, item)
+	}
+	return items
+})
+
+var jsDocTagCompletionItems = sync.OnceValue(func() []*lsproto.CompletionItem {
+	items := make([]*lsproto.CompletionItem, 0, len(jsDocTagNames))
+	for _, tagName := range jsDocTagNames {
+		item := &lsproto.CompletionItem{
+			Label:    "@" + tagName,
+			Kind:     ptrTo(lsproto.CompletionItemKindKeyword),
+			SortText: ptrTo(string(SortTextLocationPriority)),
+		}
+		items = append(items, item)
+	}
+	return items
+})
+
+func getJSDocTagNameCompletions() []*lsproto.CompletionItem {
+	return cloneItems(jsDocTagNameCompletionItems())
```

Comment:
Is cloning needed? Are they actually modified after the fact? (Just curious)

Reply:
Yes. They get always get `data` and can possibly get e.g. commit characters added.

Reply:
Yes. They get always get `data` and can possibly get e.g. commit characters added.